internal/provider/resource_secret_access.go

```diff
@@ -0,0 +1,332 @@
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		resp.Diagnostics.Append(plan.Applications.ElementsAs(ctx, &updatedSecretAccessInput.Applications, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	if noChange {
+		return
+	}
+
+	// revoke access to applications that are in the state but not in the plan
+	if applicationsInStateNotInPlan != nil {
+		err = s.client.Secrets.RevokeSecretAccess(&juju.GrantRevokeSecretAccessInput{
+			ModelName:    state.Model.ValueString(),
+			SecretId:     state.SecretId.ValueString(),
+			Applications: applicationsInStateNotInPlan.([]string),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Failed to revoke secret access", err.Error())
+			return
+		}
+	}
+
+	// grant access to applications that are in the plan but not in the state
+	if applicationsInPlanNotInState != nil {
+		err = s.client.Secrets.GrantSecretAccess(&juju.GrantRevokeSecretAccessInput{
+			ModelName:    state.Model.ValueString(),
+			SecretId:     state.SecretId.ValueString(),
+			Applications: applicationsInPlanNotInState.([]string),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Failed to grant secret access", err.Error())
+			return
+		}
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("update secret access %q", state.SecretId))
+}
+
+func (s *secretAccessResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret_access", "delete")
+		return
+	}
+
+	var state secretAccessResourceModel
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	applications := make([]string, len(state.Applications.Elements()))
+	resp.Diagnostics.Append(state.Applications.ElementsAs(ctx, &applications, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := s.client.Secrets.RevokeSecretAccess(&juju.GrantRevokeSecretAccessInput{
+		ModelName:    state.Model.ValueString(),
+		SecretId:     state.SecretId.ValueString(),
+		Applications: applications,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to revoke secret access", err.Error())
+		return
+	}
+
+	// Save empty list of applications into Terraform state
+	emptyApplicationList, errDiag := types.ListValue(types.StringType, []attr.Value{})
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Applications = emptyApplicationList
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("revoke secret access %q", state.SecretId))
+}
+
+// difference returns the difference between two slices.
+func difference(slice1, slice2 interface{}) interface{} {
```

Comment:
this can be useful for other resources' update methods as well, so it might be worth to put it in the internal/util.

Reply:
I'll put it into `helpers.go`